autogo/shadow.go

```diff
@@ -0,0 +1,158 @@
+//
+// The intended use case of this program is with the autogo_generate in //autogo:def.bzl. This rule will clone your primary workspace into an autogo workspace, and then run gazelle to generate rules for go packages.
+//
+// Usage:
+//   autogo -- <ORIGIN_DIR> <DEST_DIR>
+//
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// main calls shadowClone and exits non-zero on failure
+func main() {
+	if n := len(os.Args); n != 3 {
+		log.Printf("Expected 2 args, not %d", n)
+		log.Fatalf("Usage: %s <ORIGIN> <DEST>", filepath.Base(os.Args[0]))
+	}
+	if err := shadowClone(os.Args[1], os.Args[2]); err != nil {
+		log.Fatalf("Failed to clone %s to %s: %v", os.Args[1], os.Args[2], err)
+	}
+}
+
+// shadowClone walks origin to clone the directory structure and link files at the desination.
+func shadowClone(origin, dest string) error {
+	v := visitor{
+		origin: origin,
+		dest:   dest,
+	}
+	return filepath.Walk(v.origin, v.visit)
+}
+
+// action does something with the current file (mkdir or link)
+type action func(origin string, info os.FileInfo, dest string) error
+
+// visitor stores the origin we're cloning from and destionation we clone to.
+type visitor struct {
+	// origin is the path we read to determine what to clone
+	origin string
+	// dest is where we clone
+	dest string
+}
+
+// visit chooses and then performs the right action for the current path
+func (v visitor) visit(path string, info os.FileInfo, verr error) error {
+	act, dest, err := v.choose(path, info, verr)
+	if err != nil {
+		return err
+	}
+	if act == nil {
+		return nil
+	}
+	return act(path, info, dest)
+}
+
+// choose looks at the current path and returns the appropriate action:
+// - testdata directory => skip it and everything under it
+// - directory => create the directory
+// - go file (.go, .s) => create it at dest
+// - other files => do nothing
+func (v visitor) choose(path string, info os.FileInfo, verr error) (action, string, error) {
+	if verr != nil {
+		return nil, "", fmt.Errorf("failed to walk to %s: %v", path, verr)
+	}
+
+	// If origin is /foo/bar and we receive /foo/bar/something/special
+	// Change this to something/special
+	r, err := filepath.Rel(v.origin, path)
+	// First ensure the path is a child of origin
+	if err != nil {
+		return nil, "", fmt.Errorf("%q is not relative to %q: %v", path, v.origin, err)
+	}
+	if r == ".." || strings.HasPrefix(r, "../") {
+		return nil, "", fmt.Errorf("%q is not a child of %q", path, v.origin)
+	}
+
+	// TODO(fejta): handle symlink loops better
+	// Many repos (dep is one) have testdata folders which are
+	// either symlink loops back to some other part of the repo
+	// or else intentionally do not compile.
+	//
+	// The gazelle tool is not robust to packages containing files that fail to compile.
+	//
+	// For now just ignore these testdata folders (usually in vendored packages). This makes bazel build work at the cost of breaking bazel test
```

Comment:
`bazel test` not working is problematic...

Reply:
`bazel test` works in `k/k`. we would not be able to switch the implementation without fixing that.

Reply:
Agreed. And getting autogo to pass `k/k` tests is not yet a goal.

This is an experimental rule whose value is unproven -- and `k/k` is sure to have a bunch of edge cases. We can wait until we're more confident in the core value before investing in optimization/fixing edge cases.

In `k/k` the `bazel test //...` command only works if I also exclude a bunch of (vendor) packages.

In general `bazel test @autogo//some/random/package:all` should work but `bazel test @autogo//...` will likely include packages that hit scenarios we're not bothering to address yet.

Reply:
> In k/k the bazel test //... command only works if I also exclude a bunch of (vendor) packages.

sure, but we do run a `bazel test <more args>` on every PR with some value.

> In general bazel test @autogo//some/random/package:all should work but bazel test @autogo//... will likely include packages that hit scenarios we're not bothering to address yet.

 If we make autogo a wrapper instead we could make it work with EG `autogo test //... -//foo`?

I don't think we need to block on checking in an experiment but that might be worth considering..

Reply:
`bazel test <more args>` works just fine right now. It will work for some targets, fail others.

Agree that an `autogo` wrapper is something to explore, and that I would prefer we not block this PR on that exploration.

Reply:
it sounds like we can't do full target selection with this though eg `a - b`? not blocking at all, but that seems problematic going forward, imho.

Reply:
What does `a - b` mean? Those are not valid bazel labels.